src/compiler/scala/tools/nsc/transform/patmat/MatchCodeGen.scala

```diff
@@ -204,22 +204,41 @@ trait MatchCodeGen extends Interface {
         // res: T
         // returns MatchMonad[T]
         def one(res: Tree): Tree = matchEnd APPLY (res) // a jump to a case label is special-cased in typedApply
-        protected def zero: Tree = nextCase APPLY ()
+        protected final def zero: Tree = nextCase APPLY ()
+        override def ifThenElseZero(c: Tree, thenp: Tree): Tree = {
+          if (settings.YpatmatFlatAST.value) {
+            val z = zero
+            thenp match {
+              case If(c1, thenp1, elsep1) if z equalsStructure elsep1 =>
+                If(c AND c1, thenp1, elsep1) // cleaner, leaner trees
```

Comment:
Will this case still trigger under the new scheme? IIUC, then `thenp1` part will have been fully translated, including the reversal with `If(NOT(c1), zero, ...) :: thenp1`, before we match on `thenp` in an enclosing if/then/else. It seems to me that this case should instead recognize the shapes produced by the cases below. Shouldn't it?

Reply:
As I mention at https://github.com/scala-js/scala-js/pull/3649#issuecomment-492628466, I am a bit more convinced that my above comment is indeed relevant.

Reply:
I'm pretty sure that it is dead code under the new scheme.

Here's how the motivating test case for the #4863 changes compilation over time, up to the new scheme:

https://github.com/retronym/patmat-ast/commits/master

The previous change eliminated the nested `If`s, but did so by introducing nested `Apply` in the condition: `(c1 && c2 && ... cn)`.

The new scheme does better, AST-depth-wise, with one condition in each flattened `If`.

Reply:
This reminds me of the recent issues we've had with string interpolation AST depth of `"a1" + "a1" + ... "a256"`. My latest proposal is to add parens to reduce depth.

We could introduce an AST node to represent reductions like these in a flat way `Reduce(op: Symbol, operations: List[Tree])`, but that would require downstream compiler phases (including in plugins) to be modified to deal with it. We could probably cheat and squeeze it into an `Apply` node as well to look like a varargs call to a synthetic method, but that wouldn't capture the by-name-ness of the operands to `&&`.